### PR TITLE
[bugfix] Add reloading support using BaseFinetuning

### LINKED ISSUE #6891: [BUG] `BaseFinetuning` not working with `resume_from_checkpoint`
## 🐛 Bug

Using `BaseFinetuning` will add parameter groups during training, as a result when trying to load from checkpoint you will get the following `ValueError: loaded state dict has a different number of parameter groups`. Because when loading these groups won't exist yet !

### To Reproduce

```python
from pl_bolts.models.regression import LinearRegression
import pytorch_lightning as pl
from pl_bolts.datamodules import SklearnDataModule
from sklearn.datasets import load_boston
from pytorch_lightning.callbacks.finetuning import BaseFinetuning
from pytorch_lightning.callbacks import ModelCheckpoint

X, y = load_boston(return_X_y=True)
dm = SklearnDataModule(X, y)

class TmpLinearRegression(LinearRegression):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.to_freeze = nn.Linear(1,1)
        
    def configure_optimizers(self):
        return self.optimizer(self.linear.parameters(), lr=self.hparams.learning_rate)

class LRFinetuner(BaseFinetuning):
    def freeze_before_training(self, pl_module):
        self.freeze(modules=pl_module.to_freeze)

    def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
        if current_epoch == 1:
            self.unfreeze_and_add_param_group(
                    modules=pl_module.to_freeze,
                    optimizer=optimizer,
                    initial_denom_lr=10,
                )
    

model = TmpLinearRegression(input_dim=13)
trainer = pl.Trainer(max_epochs=2, callbacks=[ModelCheckpoint(dirpath="./tmp", save_last=True), LRFinetuner()]) 
trainer.fit(model, train_dataloader=dm.train_dataloader(), val_dataloaders=dm.val_dataloader())

model = TmpLinearRegression(input_dim=13)
trainer = pl.Trainer(max_epochs=3, resume_from_checkpoint="tmp/last.ckpt") 
trainer.fit(model, train_dataloader=dm.train_dataloader(), val_dataloaders=dm.val_dataloader())
```

Bug: 

```
/usr/local/lib/python3.8/dist-packages/pytorch_lightning/trainer/connectors/checkpoint_connector.py in restore_training_state(self, checkpoint)
    181         optimizer_states = checkpoint['optimizer_states']
    182         for optimizer, opt_state in zip(self.trainer.optimizers, optimizer_states):
--> 183             optimizer.load_state_dict(opt_state)
    184 
    185             # move optimizer to GPU 1 weight at a time

/usr/local/lib/python3.8/dist-packages/torch/optim/optimizer.py in load_state_dict(self, state_dict)
    117 
    118         if len(groups) != len(saved_groups):
--> 119             raise ValueError("loaded state dict has a different number of "
    120                              "parameter groups")
    121         param_lens = (len(g['params']) for g in groups)

ValueError: loaded state dict has a different number of parameter groups
```

### Environment
pl version : `1.2.6`

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds support for reloading the optimizer_state when training with BaseFinetuning.

Fixes #6891

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/finetuning.py [hunk 1]
```diff
@@ -82,6 +84,32 @@ def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
                     )
     """
 
+    def __init__(self):
+        self._internal_state = {}
+
+    def on_save_checkpoint(self, trainer: 'pl.Trainer', pl_module: LightningModule,
+                           checkpoint: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/finetuning.py:91]: inconsistent formatting

File: pytorch_lightning/trainer/callback_hook.py [hunk 2]
```diff
@@ -299,10 +306,16 @@ def on_load_checkpoint(self, checkpoint):
...
-                state = callback_states.get(type(callback))
-                if state:
-                    state = deepcopy(state)
-                    callback.on_load_checkpoint(state)
+                state = deepcopy(callback_states.get(type(callback)))
+                if self.__is_old_signature_on_load_checkpoint(callback.on_load_checkpoint):
+                    rank_zero_deprecation(
+                        "`Callback.on_load_checkpoint` signature has changed in v1.3."
+                        " `Trainer` and `LightningModule` parameter are been added."
+                        " Support for the old signature will be removed in v1.5"
```
Comment (awaelchli) [inline pytorch_lightning/trainer/callback_hook.py:314]:
```suggestion
                        "`Callback.on_load_checkpoint` signature has changed in v1.3."
                        " `Trainer` and `LightningModule` parameters have been added."
                        " Support for the old signature will be removed in v1.5"
```

File: pytorch_lightning/trainer/callback_hook.py [hunk 3]
```diff
@@ -299,10 +306,16 @@ def on_load_checkpoint(self, checkpoint):
         # https://github.com/pytorch/xla/issues/2773
         if callback_states is not None:
             for callback in self.callbacks:
-                state = callback_states.get(type(callback))
-                if state:
-                    state = deepcopy(state)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/callback_hook.py:304]: we shouldn't call the load checkpoint if the state is empty / not present, unless we also save empty states too. Maybe we should do that, I don't know. In any case, users have to touch all their implementations of `on_load_checkpoint`, which is not great :(

Comment (awaelchli) [reply pytorch_lightning/trainer/callback_hook.py:304]: We have other callback saving/loading related changes planned for 1.4 #6467.

File: tests/deprecated_api/test_remove_1-5.py [hunk 4]
```diff
@@ -93,6 +93,81 @@ def on_save_checkpoint(self, *args):
...
+    model = BoringModel()
+    trainer_kwargs = {
+        "default_root_dir": tmpdir,
+        "max_epochs": 3,
+    }
+    chk = ModelCheckpoint(save_last=True)
+    trainer = Trainer(
+        **trainer_kwargs, callbacks=[NewSignatureOnLoadCheckpoint(),
+                                     ValidSignature1(),
+                                     ValidSignature2(), chk]
```
Comment (awaelchli) [inline tests/deprecated_api/test_remove_1-5.py:157]: can we format with one tab? 😊

File: tests/deprecated_api/test_remove_1-5.py [hunk 5]
```diff
@@ -93,6 +93,81 @@ def on_save_checkpoint(self, *args):
...
+    trainer_kwargs = {
+        "default_root_dir": tmpdir,
+        "max_epochs": 3,
+    }
+    chk = ModelCheckpoint(save_last=True)
+    trainer = Trainer(**trainer_kwargs, callbacks=[OldSignature(), chk])
+    trainer.fit(model)
+
+    with pytest.deprecated_call(match="old signature will be removed in v1.5"):
+        trainer_kwargs["max_epochs"] = 5
```
Comment (awaelchli) [inline tests/deprecated_api/test_remove_1-5.py:130]:
```suggestion
        trainer_kwargs["max_steps"] = 1
```
one step is sufficient to trigger the warning.

File: tests/deprecated_api/test_remove_1-5.py [hunk 6]
```diff
@@ -93,6 +93,81 @@ def on_save_checkpoint(self, *args):
...
+    class OldSignature(BaseSignatureOnLoadCheckpoint):
+
+        def on_load_checkpoint(self, callback_state) -> None:
+            assert callback_state is None
+            self.on_load_checkpoint_called = True
+
+    model = BoringModel()
+    trainer_kwargs = {
+        "default_root_dir": tmpdir,
+        "max_epochs": 3,
```
Comment (awaelchli) [inline tests/deprecated_api/test_remove_1-5.py:123]:
```suggestion
        "max_steps": 3,
```

Comment (awaelchli) [reply tests/deprecated_api/test_remove_1-5.py:123]: why multiple epochs everywhere in this test?

**Review by awaelchli — COMMENTED** [review]
the signature update could also have been in a separate PR, but LGTM

The only part I am not convinced about is passing the empty state to the hooks. This requires the user to change their implementation if I understand correctly. And it also requires us to update all our callbacks.

File: pytorch_lightning/callbacks/finetuning.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:91]: Done !

File: pytorch_lightning/trainer/callback_hook.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/callback_hook.py:314]: Done !

File: tests/deprecated_api/test_remove_1-5.py (hunk 4 above)
Comment (tchaton) [reply tests/deprecated_api/test_remove_1-5.py]: Done !

File: tests/deprecated_api/test_remove_1-5.py (hunk 5 above)
Comment (tchaton) [reply tests/deprecated_api/test_remove_1-5.py:145]: Yes, should at lest 4 :)

File: tests/deprecated_api/test_remove_1-5.py (hunk 6 above)
Comment (tchaton) [reply tests/deprecated_api/test_remove_1-5.py]: Done ! thanks :)

File: pytorch_lightning/callbacks/base.py [hunk 7]
```diff
@@ -273,10 +273,19 @@ def on_save_checkpoint(
...
-    def on_load_checkpoint(self, callback_state: Dict[str, Any]) -> None:
+    def on_load_checkpoint(
+        self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule', callback_state: Dict[str, Any]
+    ) -> None:
         """Called when loading a model checkpoint, use to reload state.
 
+        .. note::
+
+            The ``on_load_checkpoint`` will be called only if a state is provided.
+            Therefore, ``on_save_checkpoint`` hook need to be overridden to return a state.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/base.py:284]:
```suggestion
        Note:
            The ``on_load_checkpoint`` will be called only if a state is provided.
            Therefore, ``on_save_checkpoint`` hook need to be overridden to return a state.
```
can the note go at the bottom? I think it's best if args docs always come first

Comment (awaelchli) [reply pytorch_lightning/callbacks/base.py:284]: Also I don't understand the second sentence:

> Therefore, ``on_save_checkpoint`` hook need to be overridden to return a state.

I think what you are trying to say relates more to special cases when the hook is used for more than just reloading state.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/callbacks/base.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:284]: Yes. Exactly.

Comment (tchaton) [reply pytorch_lightning/callbacks/base.py]: Reformulated :)

File: pytorch_lightning/callbacks/base.py [hunk 8]
```diff
@@ -273,11 +273,21 @@ def on_save_checkpoint(
...
+    ) -> None:
         """Called when loading a model checkpoint, use to reload state.
 
         Args:
+            trainer: the current Trainer instance.
+            pl_module: the current 'pl.LightningModule' instance.
             callback_state: the callback state returned by ``on_save_checkpoint``.
+
+        .. note::
+
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/base.py:287]:
```suggestion
        Note:

```
Lightning uses Google style docstrings 
https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/callbacks/base.py [hunk 9]
```diff
@@ -273,11 +273,20 @@ def on_save_checkpoint(
...
         pass
 
-    def on_load_checkpoint(self, callback_state: Dict[str, Any]) -> None:
+    def on_load_checkpoint(
+        self, trainer: 'pl.Trainer', pl_module: 'pl.LightningModule', callback_state: Dict[str, Any]
+    ) -> None:
         """Called when loading a model checkpoint, use to reload state.
 
         Args:
+            trainer: the current Trainer instance.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:282]:
```suggestion
            trainer: the current :class:`~pytorch_lightning.trainer.Trainer` instance.
```

File: pytorch_lightning/callbacks/base.py [hunk 10]
```diff
@@ -273,11 +273,20 @@ def on_save_checkpoint(
...
+            pl_module: the current 'pl.LightningModule' instance.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:283]:
```suggestion
            pl_module: the current :class:`~pytorch_lightning.core.lightning.LightningModule` instance.
```

File: pytorch_lightning/callbacks/base.py [hunk 11]
```diff
@@ -273,11 +273,20 @@ def on_save_checkpoint(
...
             callback_state: the callback state returned by ``on_save_checkpoint``.
+
+        Note:
+            The ``on_load_checkpoint`` won't be called with an undefined state.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:287]: What do you mean exactly with "an undefined state"?

File: pytorch_lightning/callbacks/finetuning.py [hunk 12]
```diff
@@ -82,6 +84,36 @@ def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
...
+        checkpoint: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        return self._internal_state
+
+    def on_load_checkpoint(
+        self, trainer: 'pl.Trainer', pl_module: LightningModule, callback_state: Dict[str, Any]
+    ) -> None:
+        self._internal_state = callback_state
+        # restore the param_groups created during the previous training.
+        _map_name_to_p = {n: p for n, p in pl_module.named_parameters()}
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:103]:
```suggestion
        _map_name_to_p = dict(pl_module.named_parameters())
```

Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:103]: Also, no point in prefixing local variables with underscore

Comment (carmocca) [reply pytorch_lightning/callbacks/finetuning.py:103]: Also, maybe rename the variable to `named_parameters`?

File: pytorch_lightning/callbacks/finetuning.py [hunk 13]
```diff
@@ -82,6 +84,36 @@ def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
...
+        for opt_idx, optimizer in enumerate(trainer.optimizers):
+            param_groups = self._param_groups_state_to_param_groups(
+                deepcopy(self._internal_state[opt_idx]), _map_name_to_p
+            )
+            optimizer.param_groups = param_groups
+
+    def _param_groups_state_to_param_groups(
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:110]: rename to `_restore_named_parameters`?

File: pytorch_lightning/callbacks/finetuning.py [hunk 14]
```diff
@@ -82,6 +84,36 @@ def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
                     )
     """
 
+    def __init__(self):
+        self._internal_state = {}
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:88]: Can you define the proper type here? If I'm not mistaken it should be something like

`Dict[int, Dict[str, Tensor]]`

File: pytorch_lightning/callbacks/finetuning.py [hunk 15]
```diff
@@ -82,6 +84,36 @@ def finetune_function(self, pl_module, current_epoch, optimizer, optimizer_idx):
...
+
+    def on_save_checkpoint(
+        self,
+        trainer: 'pl.Trainer',
+        pl_module: LightningModule,
+        checkpoint: Dict[str, Any],
+    ) -> Dict[str, Any]:
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:95]: The return type is not correct

File: pytorch_lightning/callbacks/finetuning.py [hunk 16]
```diff
@@ -234,10 +266,39 @@ def unfreeze_and_add_param_group(
...
         self.freeze_before_training(pl_module)
 
+    def _add_to_internal_state(
+        self, pl_module: LightningModule, opt_idx: int, current_param_groups: List[Dict[str, Any]]
+    ) -> None:
+        """
+        This function save the new param_group metadata inside `BaseFinetuning` Callback `internal_state`.
+        The tensors are being mapped to their names for memory optimization.
+        """
+        map_p_to_name = {p: n for n, p in pl_module.named_parameters()}
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:276]: Same comments here

File: pytorch_lightning/callbacks/finetuning.py [hunk 17]
```diff
@@ -234,10 +266,39 @@ def unfreeze_and_add_param_group(
...
+        for g in current_param_groups:
+            group_state = {k: v for k, v in g.items() if k != 'params'}
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:278]: Why do we need to skip `'params'` if we are going to set it in the next line?

File: pytorch_lightning/callbacks/finetuning.py [hunk 18]
```diff
@@ -234,10 +266,39 @@ def unfreeze_and_add_param_group(
...
+            group_state['params'] = [map_p_to_name[p] for p in g['params']]
+            self._internal_state[opt_idx].append(group_state)
+
+    def _store(
+        self, pl_module: LightningModule, opt_idx: int, len_previous_param_groups: List[Dict[str, Any]],
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:283]: `len_previous_param_groups` is an int, isnt it?

File: pytorch_lightning/callbacks/finetuning.py [hunk 19]
```diff
@@ -305,6 +366,7 @@ def __init__(
         verbose: bool = False,
         round: int = 12,
     ):
+        super().__init__()
```
Comment (carmocca) [inline pytorch_lightning/callbacks/finetuning.py:369]: Any reason why you removed this?

File: pytorch_lightning/callbacks/base.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:282]: Done !

File: pytorch_lightning/callbacks/base.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:283]: Done !

File: pytorch_lightning/callbacks/base.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/base.py:287]: `None`.

File: pytorch_lightning/callbacks/finetuning.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:103]: Done !

Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:103]: Done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:110]: Done !

File: pytorch_lightning/callbacks/finetuning.py (hunk 19 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:369]: It was missing.

File: pytorch_lightning/callbacks/finetuning.py (hunk 18 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:283]: Yes, good catch !

File: pytorch_lightning/callbacks/finetuning.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py:278]: To make it slightly cleaner for the parsing.

File: pytorch_lightning/callbacks/finetuning.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/finetuning.py]: Done !

**Review by carmocca — APPROVED** [review]

File: tests/callbacks/test_finetuning_callback.py [hunk 20]
```diff
@@ -16,15 +16,29 @@
...
 from torch.utils.data import DataLoader
 
 from pytorch_lightning import LightningModule, seed_everything, Trainer
-from pytorch_lightning.callbacks import BackboneFinetuning, BaseFinetuning
+from pytorch_lightning.callbacks import BackboneFinetuning, BaseFinetuning, ModelCheckpoint
 from pytorch_lightning.callbacks.base import Callback
 from tests.helpers import BoringModel, RandomDataset
 
 
+class TestTestBackboneFinetuningCallbackCallback(BackboneFinetuning):
```
Comment (ethanwharris) [inline tests/callbacks/test_finetuning_callback.py:28]: Is this deliberate or should it be `TestBackboneFinetuningCallback`?

**Review by ethanwharris — APPROVED** [review]
LGTM, small query

File: tests/callbacks/test_finetuning_callback.py (hunk 20 above)
Comment (tchaton) [reply tests/callbacks/test_finetuning_callback.py:28]: Wrong copy / paste ahah thanks :)

**Pull request merged**